### PR TITLE
Components logging

### PR DESCRIPTION
**Pull request by MakarenkoAI**
* [ ] Update changelog if needed
* [ ] Update SCn documentation if needed

depends on https://github.com/MksmOrlov/ostis-kb-lib/tree/test

### CODE REVIEW

File: src/manager/agents/common-module/module/keynodes/ScComponentManagerKeynodes.cpp [hunk 1]
```diff
@@ -13,14 +13,19 @@ namespace keynodes
...
 ScAddr ScComponentManagerKeynodes::concept_reusable_component_specification;
 ScAddr ScComponentManagerKeynodes::concept_complex_address;
 ScAddr ScComponentManagerKeynodes::concept_single_address;
 ScAddr ScComponentManagerKeynodes::concept_github_url;
 ScAddr ScComponentManagerKeynodes::concept_google_drive_url;
 ScAddr ScComponentManagerKeynodes::concept_need_to_install_components;
+ScAddr ScComponentManagerKeynodes::ostis_system_knowledge_base;
+ScAddr ScComponentManagerKeynodes::ostis_system_interface;
+ScAddr ScComponentManagerKeynodes::ostis_system_problem_solver;
+ScAddr ScComponentManagerKeynodes::subsystems_set;
```
Comment (MksmOrlov) [inline src/manager/agents/common-module/module/keynodes/ScComponentManagerKeynodes.cpp]:
```suggestion
ScAddr ScComponentManagerKeynodes::consept_subsystems_set;
```

File: src/manager/agents/common-module/module/keynodes/ScComponentManagerKeynodes.cpp [hunk 2]
```diff
@@ -13,14 +13,19 @@ namespace keynodes
...
+ScAddr ScComponentManagerKeynodes::concept_reusable_ui_component;
+ScAddr ScComponentManagerKeynodes::concept_reusable_embedded_ostis_system;
 ScAddr ScComponentManagerKeynodes::concept_repository;
 ScAddr ScComponentManagerKeynodes::concept_reusable_component_specification;
 ScAddr ScComponentManagerKeynodes::concept_complex_address;
 ScAddr ScComponentManagerKeynodes::concept_single_address;
 ScAddr ScComponentManagerKeynodes::concept_github_url;
 ScAddr ScComponentManagerKeynodes::concept_google_drive_url;
 ScAddr ScComponentManagerKeynodes::concept_need_to_install_components;
+ScAddr ScComponentManagerKeynodes::ostis_system_knowledge_base;
```
Comment (MksmOrlov) [inline src/manager/agents/common-module/module/keynodes/ScComponentManagerKeynodes.cpp]:
```suggestion
ScAddr ScComponentManagerKeynodes::sc_model_of_knowledge_base;
```

File: kb/system_components.gwf
Comment (MksmOrlov) [inline kb/system_components.gwf:1]: Remove this file. You should create this structure in code and check if it exists before creation. Make this resolve one time when manager starts

File: src/manager/agents/init-module/utils/sc_component_manager_command_init.cpp [hunk 3]
```diff
@@ -72,8 +73,16 @@ bool ScComponentManagerCommandInit::ProcessRepositories(ScMemoryContext * contex
...
   }
 
+  ScAddr component;
   for (ScAddr const & componentSpecificationAddr : currentComponentsSpecificationsAddrs)
   {
+    // TODO componentSpecificationAddr as specification of component
+    component = common_utils::CommonUtils::GetComponentBySpecification(*context, componentSpecificationAddr);
+    if (common_utils::CommonUtils::CheckIfInstalled(*context, component))
+    {
+      SC_LOG_DEBUG("Component \"" + context->HelperGetSystemIdtf(component) + "\" is already installed");
```
Comment (MksmOrlov) [inline src/manager/agents/init-module/utils/sc_component_manager_command_init.cpp]:
```suggestion
      SC_LOG_DEBUG("Component \"" << context->HelperGetSystemIdtf(component) << "\" is already installed");
```
Use only << to concat in SC_LOG macros. Check everywhere

File: src/manager/agents/common-module/module/utils/common_utils.cpp [hunk 4]
```diff
@@ -26,6 +27,16 @@ void CommonUtils::InitParametersMap()
...
+
+  componentsClasses = {
+      {keynodes::ScComponentManagerKeynodes::concept_reusable_ui_component,
+       keynodes::ScComponentManagerKeynodes::ostis_system_interface},
+      {keynodes::ScComponentManagerKeynodes::concept_reusable_ps_component,
+       keynodes::ScComponentManagerKeynodes::ostis_system_problem_solver},
+      {keynodes::ScComponentManagerKeynodes::concept_reusable_kb_component,
+       keynodes::ScComponentManagerKeynodes::ostis_system_knowledge_base},
+      {keynodes::ScComponentManagerKeynodes::concept_reusable_embedded_ostis_system,
+        keynodes::ScComponentManagerKeynodes::subsystems_set}};
```
Comment (MksmOrlov) [inline src/manager/agents/common-module/module/utils/common_utils.cpp]: use map instead of vector

File: kb/system_components.gwf
Comment (MksmOrlov) [reply kb/system_components.gwf:1]: You can make scs file with template (structure with variables) and find it by system identifier in code and then build and generate by template

File: src/manager/agents/common-module/module/utils/common_utils.cpp [hunk 5]
```diff
@@ -192,4 +203,71 @@ std::map<std::string, ScAddr> CommonUtils::GetElementsLinksOfSet(ScMemoryContext
...
+      continue;
+
+    ScIterator5Ptr myselfDecomposition = m_memoryCtx.Iterator5(
+        myselfAddr,
+        ScType::EdgeDCommonConst,
+        ScType::NodeConst,
+        ScType::EdgeAccessConstPosPerm,
+        keynodes::ScComponentManagerKeynodes::nrel_decomposition);
+    if (!myselfDecomposition->Next())
+      continue;
```
Comment (MksmOrlov) [inline src/manager/agents/common-module/module/utils/common_utils.cpp]: странная логика. Если не нашли декомпозицию майселфа, то идем ещё раз по циклу классов компонента и ищем опять? Но действия в цикле не влияют на это условие. В общем look at the gwf comment and remove this logic from cycle

File: src/manager/agents/common-module/module/utils/common_utils.cpp [hunk 6]
```diff
@@ -192,4 +203,71 @@ std::map<std::string, ScAddr> CommonUtils::GetElementsLinksOfSet(ScMemoryContext
...
 }
 
+ScAddr CommonUtils::GetAddrComponentOfMyself(ScMemoryContext & m_memoryCtx, ScAddr const & component)
+{
+  ScAddr myselfAddr = m_memoryCtx.HelperFindBySystemIdtf("myself");
+  ScAddr componentDecomposition;
+  if (!myselfAddr.IsValid())
+    return componentDecomposition;
+  for (std::vector<ScAddr> const & componentClass : common_utils::CommonUtils::componentsClasses)
+  {
```
Comment (MksmOrlov) [inline src/manager/agents/common-module/module/utils/common_utils.cpp]: This cycle is sooo strange. Why do you iterate over classes

File: src/manager/agents/common-module/module/utils/common_utils.cpp [hunk 7]
```diff
@@ -192,4 +203,71 @@ std::map<std::string, ScAddr> CommonUtils::GetElementsLinksOfSet(ScMemoryContext
   return elements;
 }
 
+ScAddr CommonUtils::GetAddrComponentOfMyself(ScMemoryContext & m_memoryCtx, ScAddr const & component)
```
Comment (MksmOrlov) [inline src/manager/agents/common-module/module/utils/common_utils.cpp]: addr component is not correct for English grammar. And as I see you return decomposition, not component or myself

File: src/manager/agents/common-module/module/utils/common_utils.cpp [hunk 8]
```diff
@@ -192,4 +203,71 @@ std::map<std::string, ScAddr> CommonUtils::GetElementsLinksOfSet(ScMemoryContext
...
+  ScAddr decompositionAddr = GetAddrComponentOfMyself(m_memoryCtx, component);
+  if (!decompositionAddr.IsValid() || component.IsValid())
+    return false;
+  return m_memoryCtx.HelperCheckEdge(decompositionAddr, component, ScType::EdgeAccessConstPosPerm);
+}
+
+ScAddr CommonUtils::GetComponentBySpecification(ScMemoryContext & m_memoryCtx, ScAddr const & specification)
+{
+  ScAddr component;
+  SC_LOG_DEBUG(m_memoryCtx.HelperGetSystemIdtf(specification));
```
Comment (MksmOrlov) [inline src/manager/agents/common-module/module/utils/common_utils.cpp]: this debug shouldn't be inside this method

File: src/manager/agents/common-module/module/utils/common_utils.cpp [hunk 9]
```diff
@@ -192,4 +203,71 @@ std::map<std::string, ScAddr> CommonUtils::GetElementsLinksOfSet(ScMemoryContext
...
+  ScIterator3Ptr itComponent = m_memoryCtx.Iterator3(specification, ScType::EdgeAccessConstPosPerm, ScType::NodeConst);
+  while (itComponent->Next())
+  {
+    if (m_memoryCtx.HelperCheckEdge(
+            keynodes::ScComponentManagerKeynodes::concept_reusable_component,
+            itComponent->Get(2),
+            ScType::EdgeAccessConstPosPerm))
+    {
+      return itComponent->Get(2);
+    }
```
Comment (MksmOrlov) [inline src/manager/agents/common-module/module/utils/common_utils.cpp]: We must find it onlly by rrel_key_sc_element quintuple. No need to iterate over the all specification. If there is no rrel_key_sc_element from specification to its component in all our specifications, add them.
Check sc-web, ims specifications and from MksmOrlov (cat and ostis-ui-lib). Check all our repos with specification please

**Review by MksmOrlov — CHANGES_REQUESTED** [review]
I tried to install cat and got debug that already have it. By I don't

File: kb/system_components.gwf
Comment (MksmOrlov) [reply kb/system_components.gwf:1]: Rename nrel_decomposition from myself to nrel_ostis_system_decomposition

Comment (MksmOrlov) [reply kb/system_components.gwf:1]: User interface is subsystem

File: src/manager/agents/common-module/module/utils/common_utils.cpp (hunk 4 above)
Comment (MakarenkoAI) [reply src/manager/agents/common-module/module/utils/common_utils.cpp]: https://stackoverflow.com/questions/37301160/invalid-operands-to-binary-expression-const-and-const get this issue

File: src/manager/agents/common-module/module/utils/common_utils.cpp (hunk 9 above)
Comment (MakarenkoAI) [reply src/manager/agents/common-module/module/utils/common_utils.cpp]: https://github.com/MksmOrlov/ostis-kb-lib/tree/test

File: kb/system_components.gwf
Comment (MakarenkoAI) [reply kb/system_components.gwf:1]: in main.cpp?

**Pull request closed without merging**